extension/awsmiddleware/middleware.go

```diff
@@ -0,0 +1,182 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package awsmiddleware // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/awsmiddleware"
+
+import (
+	"encoding"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/smithy-go/middleware"
+	"go.opentelemetry.io/collector/extension"
+)
+
+var (
+	errNotFound            = errors.New("middleware not found")
+	errNotMiddleware       = errors.New("extension is not an AWS middleware")
+	errInvalidHandler      = errors.New("invalid handler")
+	errUnsupportedPosition = errors.New("unsupported position")
+)
+
+// HandlerPosition is the relative position of a handler used during insertion.
+type HandlerPosition int
+
+var _ encoding.TextMarshaler = (*HandlerPosition)(nil)
+var _ encoding.TextUnmarshaler = (*HandlerPosition)(nil)
+
+const (
+	After HandlerPosition = iota
+	Before
+
+	afterStr  = "after"
+	beforeStr = "before"
+)
+
+// String returns the string representation of the position.
+// Returns an empty string if position is unsupported.
+func (h HandlerPosition) String() string {
+	switch h {
+	case Before:
+		return beforeStr
+	case After:
+		return afterStr
+	default:
+		return ""
+	}
+}
+
+// MarshalText converts the position into a string. Returns an error
+// if unsupported.
+func (h HandlerPosition) MarshalText() (text []byte, err error) {
+	s := h.String()
+	if s == "" {
+		return nil, fmt.Errorf("%w: %[2]T(%[2]d)", errUnsupportedPosition, h)
+	}
+	return []byte(h.String()), nil
+}
+
+// UnmarshalText converts the string into a position. Returns an error
+// if unsupported.
+func (h *HandlerPosition) UnmarshalText(text []byte) error {
+	switch s := string(text); s {
+	case afterStr:
+		*h = After
+	case beforeStr:
+		*h = Before
+	default:
+		return fmt.Errorf("%w: %s", errUnsupportedPosition, s)
+	}
+	return nil
+}
+
+// metadata is used to differentiate between handlers and determine
+// relative positioning within their groups.
+type metadata interface {
+	// ID must be unique. It cannot clash with existing middleware.
+	ID() string
+	// Position to insert the handler.
+	Position() HandlerPosition
+}
+
+// RequestHandler allows for custom processing of requests.
+type RequestHandler interface {
+	metadata
+	HandleRequest(r *http.Request)
```

Comment:
At a high level, do you see the usage of `*http.Request` and `*http.Response` as appropriate? This was on my mind as something that might become more clear during the implementation.

These objects are general enough that they can be applied to more than just AWS, but they're also not accounting for the other possibilities with middleware like changing default request behavior in this example: https://aws.github.io/aws-sdk-go-v2/docs/middleware/#writing-a-custom-middleware.

Some alternatives are to apply middleware to `any` or maybe some high level AWS request/response object. Then the extension implementing this interface could have type checks to apply the specific behavior it intends. What are your thoughts after spending time on this?

Reply:
I think for our usecase the `http.Request/Response` are enough. I want to avoid `any` as much as possible. It makes it hard for others to implement the interfaces and know what to expect.

The example you linked is at the initialize step. At the build and deserialize steps, it no longer has any of the parameters or inputs that would be useful. If we're not providing access to the other steps, then I don't think there's value in making it more generic.

Reply:
Makes sense, so we still have ways of extending this in the future, if doing that would even be valuable. So we're not entering a 1-way door decision.